tracer/src/Datadog.Trace/Util/SmallCacheOrNoCache.cs

```diff
@@ -0,0 +1,76 @@
+// <copyright file="SmallCacheOrNoCache.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+#nullable enable
+using System;
+using System.Collections.Concurrent;
+using System.Threading;
+using Datadog.Trace.Logging;
+
+namespace Datadog.Trace.Util;
+
+/// <summary>
+/// A wrapper around a ConcurrentDictionary that disables caching if there is too many entries.
+/// To be used in situations where we expect a low cardinality, but where uncommon setups could break that assumption
+/// </summary>
+internal class SmallCacheOrNoCache<TKey, TValue>
+    where TKey : notnull
+{
+    private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor(typeof(SmallCacheOrNoCache<TKey, TValue>));
+
+    private readonly int _maxCapacity;
+    private readonly string _logFriendlyKeyType;
+
+    private ConcurrentDictionary<TKey, TValue>? _cache = new();
+
+    /// <param name="maxCapacity">The number of entries after which we'll consider that caching is pointless</param>
+    /// <param name="logFriendlyKeyType">Only used for logging, what kind of keys are used in this cache. Use Plural.</param>
+    public SmallCacheOrNoCache(int maxCapacity, string logFriendlyKeyType)
+    {
+        _maxCapacity = maxCapacity;
+        _logFriendlyKeyType = logFriendlyKeyType;
+    }
+
+    /// <summary>
+    /// Gets a value indicating whether the cache has been disabled after receiving too many entries
```

Comment:
niptick - I think it is the inverse, it gets whether caching is enabled

```suggestion
    /// Gets a value indicating whether caching is currently enabled. Caching is disabled after receiving too many entries.
```

Reply:
ah yes you're probably right, english language issue here, I thought "whether" didn't indicate in one direction specifically :p